src/xenia/cpu/stack_walker_posix.cc

```diff
@@ -9,16 +9,293 @@
+    struct sigaction action {};
+    struct sigaction previous_action {};
+    action.sa_flags = SA_SIGINFO;
+    action.sa_sigaction = [](int signal, siginfo_t* info, void* context) {
+      std::unique_lock<std::mutex> lock(signal_handler_arg.mutex);
+      unw_getcontext(&signal_handler_arg.context_);
+      signal_handler_arg.set = true;
+      signal_handler_arg.cv.notify_one();
+    };
+    sigemptyset(&action.sa_mask);
+    sigaction(CAPTURE_SIGNAL, &action, &previous_action);
+
+    // Send signal
+    pthread_kill(reinterpret_cast<pthread_t>(thread_handle), CAPTURE_SIGNAL);
+
+    // Wait to have data back
+    unw_context_t uc;
+    {
+      std::unique_lock<std::mutex> lock(signal_handler_arg.mutex);
+      signal_handler_arg.cv.wait(lock);
+      uc = signal_handler_arg.context_;
+      signal_handler_arg.set = false;
+    }
+
+    // Restore original handler if it existed
+    sigaction(CAPTURE_SIGNAL, &previous_action, nullptr);
+
+    // Skip signal callback frame
+    ++frame_offset;
+
+    auto capture = PosixStackCapture(std::move(signal_handler_arg.context_),
+                                     frame_offset, frame_count);
+
+    for (uint32_t i = 0; i < capture.cursors_.size(); ++i) {
+      frame_host_pcs[i] = reinterpret_cast<uintptr_t>(&capture.cursors_[i]);
+    }
+    // Two identical stack traces will generate identical hash values.
+    if (out_stack_hash) {
+      *out_stack_hash = captures_.hash_function()(capture);
+    }
+
+    auto size = capture.cursors_.size();
+
+    captures_.insert(std::move(capture));
+
+    return size;
+  }
+
+  bool ResolveStack(uint64_t* frame_host_pcs, StackFrame* frames,
+                    size_t frame_count) override {
+    for (size_t i = 0; i < frame_count; ++i) {
+      auto& frame = frames[i];
+      unw_cursor_t& cursor =
+          *reinterpret_cast<unw_cursor_t*>(frame_host_pcs[i]);
+      std::memset(&frame, 0, sizeof(frame));
+      frame.host_pc = frame_host_pcs[i];
+
+      // If in the generated range, we know it's ours.
+      // if (frame.host_pc >= code_cache_min_ && frame.host_pc <
+      //     code_cache_max_) {
+      //
+      // } else {
+      //  // Host symbol, which means either emulator or system.
+      //  frame.type = StackFrame::Type::kHost;
+      // }
+      std::array<char, sizeof(frame.host_symbol.name)> host_symbol_name = {};
+      auto ok = unw_get_proc_name(&cursor, host_symbol_name.data(),
+                                  host_symbol_name.size(),
+                                  &frame.host_symbol.address);
+      switch (ok) {
+        case UNW_ESUCCESS:
+          break;
+        case UNW_EUNSPEC:
+          return false;
+        case UNW_ENOINFO:
+          return false;
+        case UNW_ENOMEM:
+          return false;
+        default:
+          return false;
+      }
+      auto demangled_host_symbol_name = demangle(host_symbol_name.data());
+      std::strncpy(frame.host_symbol.name, demangled_host_symbol_name.c_str(),
```

Comment:
`strncpy` is dangerous (doesn't insert a null terminator in case of truncation). Overall, C strings should be avoided unless it's a hard requirement to use a preallocated number of bytes for a string.